cmd/common-main.go

```diff
@@ -961,6 +956,40 @@ func handleKMSConfig() {
 			logger.Fatal(err, "Unable to initialize a connection to KES as specified by the shell environment")
 		}
 		GlobalKMS = KMS
+
+		// If no root access/secret key has been provided by the user, we generate new random
+		// access/secret key deterministically using the KMS.
+		if globalActiveCred.AccessKey == auth.DefaultCredentials.AccessKey && globalActiveCred.SecretKey == auth.DefaultSecretKey {
+			if manager, ok := GlobalKMS.(kms.KeyManager); ok {
+				aKey, err := manager.HMAC(GlobalContext, kmsConf.DefaultKeyID, []byte("root access key"))
```

Comment:
We need `logger.Info()` here to say we generate root credentials based on KMS. 

Does this completely change how root credentials can be used with MinIO? How does the client connect to set up the system for the first time?

IMO, we should only do this if the root credential based on the API call is disabled and KMS is set.

This is the correct flow. A customer would set up the cluster properly and want static credentials to work for a while. Once their setup is done, it is a lockdown stage. 

They configure root disable; now, if KMS is set, we simply use a secret credential that is never exposed instead of minioadmin, and we can suggest that the user simply remove the ROOT envs from the config file.

This can be automated in operator as well. I would call that a `lockdown` flag in operator UI or something. 

So it is a step by step process.

Reply:
> We need logger.Info() here to say we generate root credentials based on KMS.

Sure 👍 

> Does this completely change how root credentials can be used with MinIO? How does the client connect to set up the system for the first time?

The KMS credentials are still "static" as long as the `MINIO_KMS_KES_KEY_NAME` stays the same. They are just not predictable unless how have access to the KMS. So, they don't change. For first access, the user would either need to obtain them via an API call / command from the KMS or look at the console output.

We could restrict is it to be only effective in lock-down mode. However, I don't see any problem with the current behavior or am I missing something? It's not like the credentials keep rotating or depend on any state at MinIO (apart from the KMS key name). Nevertheless if you prefer to only use KMS generated access/secret keys in lock-down mode (as of now) I can change it.

Reply:
For example, with the following env. vars MinIO starts like this:
```
```
MINIO_KMS_KES_ENDPOINT=https://127.0.0.1:7373
MINIO_KMS_KES_KEY_NAME=my-key
MINIO_KMS_KES_CAPATH=/tmp/public.crt
MINIO_KMS_KES_API_KEY=kes:v1:AD9E7FSYWrMD+VjhI6q545cYT9YOyFxZb7UnjEepYDRc
```
minio server /tmp/1
Automatically generated root access key and secret key with the KMS
Formatting 1st pool, 1 set(s), 1 drives per set.
WARNING: Host local has all drives of set. A host failure will result in data becoming unavailable.
MinIO Object Storage Server
Copyright: 2015-0000 MinIO, Inc.
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Version: DEVELOPMENT.GOGET (go1.22.0 darwin/arm64)

Status:         1 Online, 0 Offline.
S3-API: http://192.168.188.79:9000  http://127.0.0.1:9000
RootUser: P63ITYI6L07GY1VKY3ET
RootPass: 4+yRbGik+i+Yf98T0Jrw+8XtlBAtWjYznRvd0kkH+uE

Console: http://192.168.188.79:53656 http://127.0.0.1:53656
RootUser: P63ITYI6L07GY1VKY3ET
RootPass: 4+yRbGik+i+Yf98T0Jrw+8XtlBAtWjYznRvd0kkH+uE

Command-line: https://min.io/docs/minio/linux/reference/minio-mc.html#quickstart
   $ mc alias set 'myminio' 'http://192.168.188.79:9000' 'P63ITYI6L07GY1VKY3ET' '4+yRbGik+i+Yf98T0Jrw+8XtlBAtWjYznRvd0kkH+uE'

Documentation: https://min.io/docs/minio/linux/index.html
Warning: The standard parity is set to 0. This can lead to data loss.
```

The root credentials remain to be `P63ITYI6L07GY1VKY3ET` and `4+yRbGik+i+Yf98T0Jrw+8XtlBAtWjYznRvd0kkH+uE` unless you change `MINIO_KMS_KES_KEY_NAME` to some other key.

Reply:
Yeah @aead but this only visible in stdout. But not as a service or inside k8s pods.

We hide credentials there.

Reply:
I think with this we will still have the issue of users complaining about plaintext output _somewhere_ in the system. 

What about using `mc` ? our assumption here is for this to work, we need the API KEY + KEY NAME. So if you specify both to mc as part of some call (`mc admin user get-root ALIAS APIKEY KEYNAME`) then it's up to the user to appropriately sanitize bash history

and at that point it goes in the `mc` json which is also plaintext so :)

Reply:
I think we should not ever automatically do this, we should convert this into a deployment process.

Basically we need to ask users to make sure root is disabled access and KMS is configured to make it "fully secure" i.e no credentials are ever written anywhere.

Reply:
So basically:

1. Start MinIO normally
2. Create your non-root superuser
3. Ensure KMS is configured
4. Disable root API access

At this point MinIO will auto-generate a new root credential that is completely opaque to the end user, and cannot be found *anywhere*.

is that about right?

Reply:
Okay, I change the PR to only generate root credentials with the KMS if root API access is disabled.

Reply:
Updated. Now, MinIO generates the root credentials with the KMS only in lockdown mode. Now, it also does not print root access/secret key.

Reply:
Yes this is perfect. This is what we can bring in Operator mode where they can enable this after everything has been configured as per their liking and no more root credentials are needed.

Reply:
I would test this out with site replicated setup though as there is one challenge there I want to sort this out 

// cc @poornas 

Can you give this PR a test on site replication? IAM assets for sts have root credentials requirement to be replicated. So have to check how to do this PR safely.

Reply:
Now that site replication PR is merged we can rebase this PR @aead and get this ready to be merged.

Reply:
Done @harshavardhana